### PR TITLE
support traffic settings for shadow deployment with istio

### LINKED ISSUE #3772: Support Traffic Settings to SeldonDeployment  For Shadow Deployment with istio
When deploying **shadow** deployment with **istio**, we can  config the mirror deploy's Host,Subset,Port in created virtualservice .
it's very helpful to have a traffic setting which decide  mirror a **fraction** of the traffic, instead of mirroring **all requests**

### PR DESCRIPTION
**Pull request by ChowXu**
**What this PR does / why we need it**:   support traffic settings for shadow deployment with istio to meet more situations

**Special notes for your reviewer**: 

**Which issue(s) this PR fixes**:  Fixes #3772 

**Does this PR introduce a user-facing change?**:  NONE

```release-note
```

### CODE REVIEW

**Comment by seldondev** [comment]
Hi @ChowXu. Thanks for your PR.

I'm waiting for a [SeldonIO](https://github.com/orgs/SeldonIO/people) or [todo](https://github.com/orgs/todo/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/todo/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **axsaucedo**
You can assign the PR to them by writing `/assign @axsaucedo` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by cliveseldon** [comment]
Thanks @ChowXu 
I think we would also need to modify the validating webhook check for traffic to ignore the shadow value in checking traffic sums to 100. We might also want to check that the shadow traffic percentrage is >=0 <= 100.

**Comment by ChowXu** [comment]
> Thanks @ChowXu I think we would also need to modify the validating webhook check for traffic to ignore the shadow value in checking traffic sums to 100. We might also want to check that the shadow traffic percentrage is >=0 <= 100.
> 
> https://github.com/SeldonIO/seldon-core/blob/ce9a2842a95a3212044542f78aa299c26fe1264d/operator/apis/machinelearning.seldon.io/v1/seldondeployment_webhook.go#L126-L149
Thanks @cliveseldon ,  the checkTraffic's logical has been updated

File: operator/apis/machinelearning.seldon.io/v1/seldondeployment_webhook.go [hunk 1]
```diff
@@ -135,14 +135,19 @@ func checkTraffic(spec *SeldonDeploymentSpec, fldPath *field.Path, allErrs field
...
+			}
 		}
 	}
 
-	if trafficSum != 100 && (len(spec.Predictors)-shadows) > 1 {
-		allErrs = append(allErrs, field.Invalid(fldPath, spec.Predictors[0].Name, "Traffic must sum to 100 for multiple predictors"))
-	}
-	if trafficSum > 0 && trafficSum < 100 && len(spec.Predictors) == 1 {
-		allErrs = append(allErrs, field.Invalid(fldPath, spec.Predictors[0].Name, "Traffic must sum be 100 for a single predictor when set"))
+	if shadows == 0 {
```
Comment (cliveseldon) [inline operator/apis/machinelearning.seldon.io/v1/seldondeployment_webhook.go:144]: I think we need to do the checks even if shadows are present. We just need to ignore the shadow percentage in the calculation.

**Comment by axsaucedo** [comment]
Looks good - following up on some of the requests, it would be great if you could also add some docs, and potentially extend some of the istio docs to show an example (i.e. the istio notebook)

**Comment by ChowXu** [comment]
thanks @axsaucedo , i have added the shadow deploy doc upder examples.istio , and also with a tip about traffic settings usage.

**Comment by cliveseldon** [comment]
@ChowXu There are some errors compiling the operator.

**Comment by ChowXu** [comment]
thanks @cliveseldon , fix it now

File: operator/apis/machinelearning.seldon.io/v1/seldondeployment_webhook.go [hunk 2]
```diff
@@ -128,13 +128,17 @@ func checkTraffic(spec *SeldonDeploymentSpec, fldPath *field.Path, allErrs field
...
 			shadows += 1
 			if shadows > 1 {
 				allErrs = append(allErrs, field.Invalid(fldPath, spec.Predictors[i].Name, "Multiple shadows are not allowed"))
 			}
+			if p.Traffic < 0 || p.Traffic > 100 {
+				allErrs = append(allErrs, field.Invalid(fldPath, spec.Predictors[i].Name, "shadow traffic is illegal, the traffic number should be between [0, 100]"))
+			}
+		} else {
+			trafficSum = trafficSum + p.Traffic
 		}
```
Comment (cliveseldon) [inline operator/apis/machinelearning.seldon.io/v1/seldondeployment_webhook.go:142]: Can we add a go test for this?

Comment (ChowXu) [reply operator/apis/machinelearning.seldon.io/v1/seldondeployment_webhook.go:142]: sure, i have add a UT for shadow traffic validation

**Comment by seldondev** [comment]
@ChowXu: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 02cc652a719a36de40554d50f3311b9949910357 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-3780-integration-97lqk) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@ChowXu: Cannot trigger testing until a trusted user reviews the PR and leaves an `/ok-to-test` message.

<details>

In response to [this](https://github.com/SeldonIO/seldon-core/pull/3780#issuecomment-1050607044):

>/test integration


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

**Comment by seldondev** [comment]
@ChowXu: Cannot trigger testing until a trusted user reviews the PR and leaves an `/ok-to-test` message.

<details>

In response to [this](https://github.com/SeldonIO/seldon-core/pull/3780#issuecomment-1050607641):

>/retest


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

**Comment by ChowXu** [comment]
Hi @axsaucedo the integration test meet some problems, but i have not authorization to visit [pipelinerun page](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-3780-integration-97lqk) to see detail about what happended and i cannot rerun the test , could u give me a help?

**Comment by cliveseldon** [comment]
@ChowXu will run again as failed early so need to see if real issue

**Comment by seldondev** [comment]
@ChowXu: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 02cc652a719a36de40554d50f3311b9949910357 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-3780-integration-9q88g) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@ChowXu: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 02cc652a719a36de40554d50f3311b9949910357 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-3780-integration-96xqp) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Pull request merged**